src/reflect/scala/reflect/internal/Types.scala

```diff
@@ -1632,14 +1632,17 @@ trait Types
     private def normalizeImpl = {
       // TODO see comments around def intersectionType and def merge
       // scala/bug#8575 The dealias is needed here to keep subtyping transitive, example in run/t8575b.scala
-      def flatten(tps: List[Type]): List[Type] = {
+      val flattened: List[Type] = {
+        @inline
         def dealiasRefinement(tp: Type) = if (tp.dealias.isInstanceOf[RefinedType]) tp.dealias else tp
-        tps map dealiasRefinement flatMap {
-          case RefinedType(parents, ds) if ds.isEmpty => flatten(parents)
-          case tp => List(tp)
+        val buf: ListBuffer[Type] = ListBuffer.empty[Type]
+        def loop(tp: Type): Unit = dealiasRefinement(tp) match {
+          case RefinedType(parents, ds) if ds.isEmpty => parents.foreach(loop)
+          case tp => if (buf contains tp) () else buf += tp
         }
+        parents foreach loop
+        buf.toList
       }
```

Comment:
Maybe use a `mutable.LinkedHashSet` as the buffer, rather than using the `O(N)` ListBuffer.contains` inside the loop?

Below, you could replace `flattened != parents` with `!flattenedLinkedHashSet.sameElements(parents)`, and defer creation of the actuaion `List[Type]` until needed.

Reply:
I thought about using a different data structure, but I was not sure if preserving order of insertion was critical for this part of the code to work.

Reply:
The problem I see with using a `LinkedHashSet` is that, if the `flattened` list has too few elements, is more allocated space and objects than having the `LinkedList`. 

In other places and for a similar problem, in which we need to decide _containment_ for a potentially large but probably often small list of elements, I have considered introducing a little simple bloom filter. This would consist of a single `Long` variable in the stack (so no extra heap allocations) and for every element we would mark one of the bits between 0 and 63. Of course, having statistics about the actual sizes of these lists would help to make a better decision.

Reply:
Would it be too problematic to make this `LinkedHashSet` as a sort of _constant_ variable, pre-allocated and stored in a companion object, so that instead fo 

Of course, this would mean introducing a form of _global state_, even if its visibility is restricted to this method, which would impede making concurrent calls to this method. However, I am not entirely sure if the compiler has any parallelism enabled?

Reply:
This method isn't called concurrently, but it _may_ be re-entrant (ie, call itself recursively, perhaps through `.dealias`).

We do have a convenient way to use a cache for level 0 of recursion of such cases:  `ReusableInstance`.

Reply:
Ok. I have rewritten the code to use a `LinkedHashSet`, as required.

Reply:
Ok. I think that the optimisation of using a cached instance can be left for another PR, if needed.